astropy/coordinates/sky_coordinate_parsers.py

```diff
@@ -535,24 +535,11 @@ def _parse_coordinate_arg(coords, frame, units, init_kwargs):
                 skycoord_kwargs[fattrnm] = getattr(scs[0], fattrnm)
 
             # Now combine the values, to be used below
-            values = []
-            for data_attr_name, repr_attr_name in zip(
-                frame_attr_names, repr_attr_names
-            ):
-                if allunitsphrepr and repr_attr_name == "distance":
-                    # if they are *all* UnitSpherical, don't give a distance
-                    continue
-                data_vals = []
-                for sc in scs:
-                    data_val = getattr(sc, data_attr_name)
-                    data_vals.append(
-                        data_val.reshape((1,)) if sc.isscalar else data_val
-                    )
-                concat_vals = np.concatenate(data_vals)
-                # Hack because np.concatenate doesn't fully work with Quantity
-                if isinstance(concat_vals, u.Quantity):
-                    concat_vals._unit = data_val.unit
-                values.append(concat_vals)
+            values = [
+                np.concatenate([np.atleast_1d(getattr(sc, data_attr)) for sc in scs])
+                for data_attr, repr_attr in zip(frame_attr_names, repr_attr_names)
+                if not allunitsphrepr or repr_attr != "distance"
```

Comment:
Maybe, just for clarity, `not (allunitsphrepr and repr_attr == "distance")`?

Reply:
I didn't like that in the version of the code you reviewed it defined `allunitsphrepr` but actually used `not allunitsphrepr`. The newly updated code defines `not_unit_sphere` that does not have to be negated in the list comprehension.